firewood/src/merkle/stream.rs

```diff
@@ -7,399 +7,449 @@ use crate::{
+        }
+    }
+}
 
-                    let next_result = value.map(|value| {
-                        let value = value.to_vec();
+/// Returns the initial state for an iterator over the given `merkle` with root `root_node`
+/// which starts at `key`.
+fn get_iterator_intial_state<'a, S: ShaleStore<Node> + Send + Sync, T>(
+    merkle: &'a Merkle<S, T>,
+    root_node: DiskAddress,
+    key: &[u8],
+) -> Result<MerkleNodeStreamState<'a>, api::Error> {
+    // Invariant: `node`'s key is a prefix of `key`.
+    let mut node = merkle
+        .get_node(root_node)
+        .map_err(|e| api::Error::InternalError(Box::new(e)))?;
 
-                        Ok((std::mem::take(key), value))
-                    });
+    // Invariant: [matched_key_nibbles] is the key of `node` at the start
+    // of each loop iteration.
+    let mut matched_key_nibbles = vec![];
 
-                    visited_node_path.push((found_node, 0));
+    let mut unmatched_key_nibbles = Nibbles::<1>::new(key).into_iter();
 
-                    self.key_state = IteratorState::Iterating {
-                        check_child_nibble,
-                        visited_node_path,
-                    };
+    let mut iter_stack: Vec<IterationNode> = vec![];
 
-                    return Poll::Ready(next_result);
+    loop {
+        let Some(nib) = unmatched_key_nibbles.next() else {
+            // The invariant tells us `node` is a prefix of `key`.
+            // There is no more `key` left so `node` must be at `key`.
+            // Visit and return `node` first.
+            match &node.inner {
+                NodeType::Branch(_) | NodeType::Leaf(_) => {
+                    iter_stack.push(IterationNode::Unvisited {
+                        key: matched_key_nibbles.clone().into_boxed_slice(),
+                        node,
+                    });
                 }
-
-                let found_key = nibble_iter_from_parents(&visited_node_path);
-                let found_key = key_from_nibble_iter(found_key);
-
-                if found_key > *key {
-                    check_child_nibble = false;
-                    visited_node_path.pop();
+                NodeType::Extension(_) => {
+                    panic!("extension nodes shouldn't exist")
                 }
+            }
 
-                self.key_state = IteratorState::Iterating {
-                    check_child_nibble,
-                    visited_node_path,
+            return Ok(MerkleNodeStreamState::Initialized { iter_stack });
+        };
+        // `nib` is the first nibble after [matched_key_nibbles].
+
+        match &node.inner {
+            NodeType::Branch(branch) => {
+                // The next nibble in `key` is `nib`,
+                // so all children of `node` with a position > `nib`
+                // should be visited since they are after `key`.
+                iter_stack.push(IterationNode::Visited {
+                    key: matched_key_nibbles.clone().into_boxed_slice(),
```

Comment:
For a follow up PR:
I think it might make sense to do the following:
```suggestion
                    key: matched_key_nibbles.iter().copied().collect(),
```
I think you did this elsewhere in the PR too. The reason that I would suggest to do this instead of calling `clone` for turning a `Vec` into a `Box<[u8]>`, is that there are no guarantees that `Vec::clone` doesn't also create a new `Vec` with the same capacity, then turn it into a `Box<[u8]>`, dropping the extra capacity. 

In practice, that's not how the current version of `Vec::clone` works, but the only thing that's guaranteed is the API between Rust versions. 

I'm sure that you get the exact same output assembly, but I think it's generally better to be explicit about size (by using an iterator).

Reply:
Another good choice is: `key: Box::from(matched_key_nibbles.as_slice()),` since [Box::from can take a slice](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from-20)